File: Medium/Anti Diagonal Traversal of Matrix/anti-diagonal-traversal-of-matrix.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
public:
    std::vector<int> antiDiagonalPattern(std::vector<std::vector<int>>& matrix) {
        int num_rows = matrix.size();
        int num_cols = matrix[0].size();

        // Dictionary to store elements grouped by their anti-diagonal index
        std::unordered_map<int, std::vector<int>> anti_diagonal_groups;

        // Populate anti_diagonal_groups dictionary
        for (int i = 0; i < num_rows; ++i) {
            for (int j = 0; j < num_cols; ++j) {
                int anti_diagonal_index = i + j;
                anti_diagonal_groups[anti_diagonal_index].push_back(matrix[i][j]);
            }
        }

        std::vector<int> result;

        // Flatten the anti-diagonal groups into the final result list
        for (int i = 0; i < num_rows + num_cols - 1; ++i) {
            result.insert(result.end(), anti_diagonal_groups[i].begin(), anti_diagonal_groups[i].end());
        }

        return result;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;

    while (t--) {
        int n;
        cin >> n;
        vector<vector<int>> matrix;
        matrix.resize(n, vector<int>(n));

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cin >> matrix[i][j];
            }
        }

        Solution ob;
        vector<int> ans = ob.antiDiagonalPattern(matrix);
        for (int i = 0; i < ans.size(); ++i) cout << ans[i] << " ";
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Medium/Anti Diagonal Traversal of Matrix/anti-diagonal-traversal-of-matrix.py
#User function Template for python3

class Solution:
    def antiDiagonalPattern(self, matrix):
        num_rows, num_cols = len(matrix), len(matrix[0])

        # Dictionary to store elements grouped by their anti-diagonal index
        anti_diagonal_groups = {}

        # Populate anti_diagonal_groups dictionary
        for i in range(num_rows):
            for j in range(num_cols):
                anti_diagonal_index = i + j
                anti_diagonal_groups[anti_diagonal_index] = anti_diagonal_groups.get(anti_diagonal_index, []) + [matrix[i][j]]

        result = []

        # Flatten the anti-diagonal groups into the final result list
        for i in range(num_rows + num_cols - 1):
            result.extend(anti_diagonal_groups[i])

        return result


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n = int(input()) 
        inp=list(map(int,input().split()))
        matrix=[]
        k = 0
        for i in range(n):
            row = []
            for j in range(n):
                row.append(inp[k])
                k += 1
            matrix.append(row)
        ob = Solution()
        ans = ob.antiDiagonalPattern(matrix)
        for i in ans:
            print(i,end=" ")
        print()


# } Driver Code Ends